File: master-server/raft/raft_fsm_leader.go
package raft

import (
	"fmt"
	"sort"
	"time"

	"master-server/raft/logger"
	"master-server/raft/proto"
	"master-server/raft/util"
)

func (r *raftFsm) becomeLeader() {
	if r.state == stateFollower {
		panic(AppPanicError(fmt.Sprintf("[raft->becomeLeader][%v] invalid transition [follower -> leader].", r.id)))
	}

	lasti := r.raftLog.lastIndex()
	r.step = stepLeader
	r.reset(r.term, lasti, true)
	r.tick = r.tickHeartbeat
	r.leader = r.config.NodeID
	r.state = stateLeader
	r.acks = nil

	ents, err := r.raftLog.entries(r.raftLog.committed+1, noLimit)
	if err != nil {
		errMsg := fmt.Sprintf("[raft->becomeLeader][%v] unexpected error getting uncommitted entries (%v).", r.id, err)
		logger.Error(errMsg)
		panic(AppPanicError(errMsg))
	}
	nconf := numOfPendingConf(ents)
	if nconf > 1 {
		panic(AppPanicError(fmt.Sprintf("[raft->becomeLeader][%v] unexpected double uncommitted config entry.", r.id)))
	}
	if nconf == 1 {
		r.pendingConf = true
	}

	r.appendEntry(&proto.Entry{Term: r.term, Index: lasti + 1, Data: nil})
	if logger.IsEnableInfo() {
		logger.Info("raft[%v] became leader at term %d.", r.id, r.term)
	}
}

func stepLeader(r *raftFsm, m *proto.Message) {
	// These message types do not require any progress for m.From.
	switch m.Type {
	case proto.LocalMsgProp:
		if _, ok := r.replicas[r.config.NodeID]; !ok || len(m.Entries) == 0 {
			return
		}

		for i, e := range m.Entries {
			if e.Type == proto.EntryConfChange {
				if r.pendingConf {
					m.Entries[i] = &proto.Entry{Term: e.Term, Index: e.Index, Type: proto.EntryNormal}
				}
				r.pendingConf = true
			}
		}
		r.appendEntry(m.Entries...)
		r.bcastAppend()
		proto.ReturnMessage(m)
		return

	case proto.ReqMsgVote:
		if logger.IsEnableInfo() {
			logger.Info("raft[%v] stepLeader [logterm: %d, index: %d, vote: %v] rejected vote from %v [logterm: %d, index: %d] at term %d",
				r.id, r.raftLog.lastTerm(), r.raftLog.lastIndex(), r.vote, m.From, m.LogTerm, m.Index, r.term)
		}
		nmsg := proto.GetMessage()
		nmsg.Type = proto.RespMsgVote
		nmsg.To = m.From
		nmsg.Reject = true
		r.send(nmsg)
		proto.ReturnMessage(m)
		return
	}

	// All other message types require a progress for m.From (pr).
	pr, prOk := r.replicas[m.From]
	if !prOk {
		if logger.IsEnableDebug() {
			logger.Debug("raft[%v] stepLeader no progress available for %v.", r.id, m.From)
		}
		return
	}
	switch m.Type {
	case proto.RespMsgAppend:
		pr.active = true
		pr.lastActive = time.Now()

		if m.Reject {
			if logger.IsEnableDebug() {
				logger.Debug("raft[%v] received msgApp rejection(lastindex: %d) from %v for index %d.", r.id, m.RejectIndex, m.From, m.Index)
			}
			if pr.maybeDecrTo(m.Index, m.RejectIndex, m.Commit) {
				if pr.state == replicaStateReplicate {
					pr.becomeProbe()
				}
				r.sendAppend(m.From)
			}
		} else {
			oldPaused := pr.isPaused()
			if pr.maybeUpdate(m.Index, m.Commit) {
				switch {
				case pr.state == replicaStateProbe:
					pr.becomeReplicate()
				case pr.state == replicaStateSnapshot && pr.needSnapshotAbort():
					if logger.IsEnableWarn() {
						logger.Warn("raft[%v] snapshot aborted, resumed sending replication messages to %v.", r.id, m.From)
					}
					pr.becomeProbe()
				case pr.state == replicaStateReplicate:
					pr.inflight.freeTo(m.Index)
				}

				if r.maybeCommit() {
					r.bcastAppend()
				} else if oldPaused {
					r.sendAppend(m.From)
				}
			}
		}
		proto.ReturnMessage(m)
		return

	case proto.RespMsgHeartBeat:
		if pr.state == replicaStateReplicate && pr.inflight.full() {
			pr.inflight.freeFirstOne()
		}
		if !pr.pending && (pr.match < r.raftLog.lastIndex() || pr.committed < r.raftLog.committed) {
			r.sendAppend(m.From)
		}

		pr.active = true
		pr.lastActive = time.Now()
		if pr.state != replicaStateSnapshot {
			pr.pending = false
		}
		return

	case proto.LeaseMsgOffline:
		for id := range r.replicas {
			if id == r.config.NodeID {
				continue
			}
			nmsg := proto.GetMessage()
			nmsg.Type = proto.LeaseMsgTimeout
			nmsg.To = id
			r.send(nmsg)
		}
		r.becomeFollower(r.term, NoLeader)
		proto.ReturnMessage(m)
		return

	case proto.RespMsgSnapShot:
		if pr.state != replicaStateSnapshot {
			return
		}

		if m.Reject {
			if logger.IsEnableWarn() {
				logger.Warn("raft[%v] send snapshot to [%v] failed.", r.id, m.From)
			}
			pr.snapshotFailure()
			pr.becomeProbe()
		} else {
			pr.active = true
			pr.lastActive = time.Now()
			pr.becomeProbe()
			if logger.IsEnableWarn() {
				logger.Warn("raft[%v] send snapshot to [%v] succeeded, resumed replication [%s]", r.id, m.From, pr)
			}
		}

		// If snapshot finish, wait for the RespMsgAppend from the remote node before sending out the next ReqMsgAppend.
		// If snapshot failure, wait for a heartbeat interval before next try.
		pr.pause()
		proto.ReturnMessage(m)
		return
	}
}

func (r *raftFsm) becomeElectionAck() {
	r.acks = make(map[uint64]bool)
	r.acks[r.config.NodeID] = true
	if len(r.acks) >= r.quorum() {
		r.becomeLeader()
		return
	}

	r.step = stepElectionAck
	r.reset(r.term, 0, false)
	r.tick = r.tickElectionAck
	r.state = stateElectionACK
	for id := range r.replicas {
		if id == r.config.NodeID {
			continue
		}

		m := proto.GetMessage()
		m.Type = proto.ReqMsgElectAck
		m.To = id
		r.send(m)
	}
}

func stepElectionAck(r *raftFsm, m *proto.Message) {
	switch m.Type {
	case proto.LocalMsgProp:
		if logger.IsEnableDebug() {
			logger.Debug("raft[%v] no leader at term %d; dropping proposal", r.id, r.term)
		}
		proto.ReturnMessage(m)
		return

	case proto.ReqMsgAppend:
		r.becomeFollower(r.term, m.From)
		r.handleAppendEntries(m)
		proto.ReturnMessage(m)
		return

	case proto.ReqMsgHeartBeat:
		r.becomeFollower(r.term, m.From)
		return

	case proto.ReqMsgElectAck:
		r.becomeFollower(r.term, m.From)
		nmsg := proto.GetMessage()
		nmsg.Type = proto.RespMsgElectAck
		nmsg.To = m.From
		r.send(nmsg)
		proto.ReturnMessage(m)
		return

	case proto.ReqMsgVote:
		nmsg := proto.GetMessage()
		nmsg.Type = proto.RespMsgVote
		nmsg.To = m.From
		nmsg.Reject = true
		r.send(nmsg)
		proto.ReturnMessage(m)
		return

	case proto.RespMsgElectAck:
		r.replicas[m.From].active = true
		r.replicas[m.From].lastActive = time.Now()
		r.acks[m.From] = true
		if len(r.acks) >= r.quorum() {
			r.becomeLeader()
			r.bcastAppend()
		}
		proto.ReturnMessage(m)
		return
	}
}

func (r *raftFsm) tickHeartbeat() {
	r.heartbeatElapsed++
	r.electionElapsed++
	if r.pastElectionTimeout() {
		r.electionElapsed = 0
		if r.config.LeaseCheck && !r.checkLeaderLease() {
			if logger.IsEnableWarn() {
				logger.Warn("raft[%v] stepped down to follower since quorum is not active.", r.id)
			}
			r.becomeFollower(r.term, NoLeader)
		}
	}

	if r.state != stateLeader {
		return
	}

	if r.heartbeatElapsed >= r.config.HeartbeatTick {
		r.heartbeatElapsed = 0
		for id := range r.replicas {
			if id == r.config.NodeID {
				continue
			}
			if r.replicas[id].state != replicaStateSnapshot {
				r.replicas[id].resume()
			}
		}
	}
}

func (r *raftFsm) tickElectionAck() {
	r.electionElapsed++
	if r.electionElapsed >= r.config.ElectionTick {
		r.electionElapsed = 0

		m := proto.GetMessage()
		m.Type = proto.LocalMsgHup
		m.From = r.config.NodeID
		r.Step(m)
	}
}

func (r *raftFsm) checkLeaderLease() bool {
	var act int
	for id := range r.replicas {
		if id == r.config.NodeID || r.replicas[id].state == replicaStateSnapshot {
			act++
			continue
		}

		if r.replicas[id].active {
			act++
		}
		r.replicas[id].active = false
	}

	return act >= r.quorum()
}

func (r *raftFsm) maybeCommit() bool {
	mis := make(util.Uint64Slice, 0, len(r.replicas))
	for _, rp := range r.replicas {
		mis = append(mis, rp.match)
	}
	sort.Sort(sort.Reverse(mis))
	mci := mis[r.quorum()-1]
	isCommit := r.raftLog.maybeCommit(mci, r.term)
	if r.state == stateLeader && r.replicas[r.config.NodeID] != nil {
		r.replicas[r.config.NodeID].committed = r.raftLog.committed
	}
	return isCommit
}

func (r *raftFsm) bcastAppend() {
	for id := range r.replicas {
		if id == r.config.NodeID {
			continue
		}
		r.sendAppend(id)
	}
}

func (r *raftFsm) sendAppend(to uint64) {
	pr := r.replicas[to]
	if pr.isPaused() {
		return
	}

	var (
		term       uint64
		ents       []*proto.Entry
		errt, erre error
		m          *proto.Message
	)
	fi := r.raftLog.firstIndex()
	if pr.next >= fi {
		term, errt = r.raftLog.term(pr.next - 1)
		ents, erre = r.raftLog.entries(pr.next, r.config.MaxSizePerMsg)
	}
	if pr.next < fi || errt != nil || erre != nil {
		if !pr.active {
			if logger.IsEnableDebug() {
				logger.Debug("raft[%v] sendAppend ignore sending snapshot to %v since it is not recently active.", r.id, to)
			}
			return
		}

		snapshot, err := r.sm.Snapshot()
		if err != nil || snapshot.ApplyIndex() < fi-1 {
			panic(AppPanicError(fmt.Sprintf("[raft->sendAppend][%v]failed to send snapshot[%d] to %v because snapshot is unavailable, error is: \r\n%v", r.id, snapshot.ApplyIndex(), to, err)))
		}

		m = proto.GetMessage()
		m.Type = proto.ReqMsgSnapShot
		m.To = to
		m.Snapshot = snapshot
		snapMeta := proto.SnapshotMeta{Index: snapshot.ApplyIndex(), Peers: make([]proto.Peer, 0, len(r.replicas))}
		if snapTerm, err := r.raftLog.term(snapMeta.Index); err != nil {
			panic(AppPanicError(fmt.Sprintf("[raft->sendAppend][%v]failed to send snapshot to %v because snapshot is unavailable, error is: \r\n%v", r.id, to, err)))
		} else {
			snapMeta.Term = snapTerm
		}
		for _, p := range r.replicas {
			snapMeta.Peers = append(snapMeta.Peers, p.peer)
		}
		m.SnapshotMeta = snapMeta
		pr.becomeSnapshot(snapMeta.Index)

		if logger.IsEnableDebug() {
			logger.Debug("raft[%v] sendAppend [firstindex: %d, commit: %d] sent snapshot[index: %d, term: %d] to [%v][%s]",
				r.id, fi, r.raftLog.committed, snapMeta.Index, snapMeta.Term, to, pr)
		}
	} else {
		m = proto.GetMessage()
		m.Type = proto.ReqMsgAppend
		m.To = to
		m.Index = pr.next - 1
		m.LogTerm = term
		m.Commit = r.raftLog.committed
		m.Entries = append(m.Entries, ents...)

		if n := len(m.Entries); n != 0 {
			switch pr.state {
			case replicaStateReplicate:
				last := m.Entries[n-1].Index
				pr.update(last)
				pr.inflight.add(last)
			case replicaStateProbe:
				pr.pause()
			default:
				errMsg := fmt.Sprintf("[repl->sendAppend][%v] is sending append in unhandled state %s.", r.id, pr.state)
				logger.Error(errMsg)
				panic(AppPanicError(errMsg))
			}
		}
	}
	pr.pending = true
	r.send(m)
}

func (r *raftFsm) appendEntry(es ...*proto.Entry) {
	r.raftLog.append(es...)
	r.replicas[r.config.NodeID].maybeUpdate(r.raftLog.lastIndex(), r.raftLog.committed)
	r.maybeCommit()
}

File: data-server/src/raft/src/impl/raft_fsm_leader.cpp
#include "raft_fsm.h"

#include <algorithm>
#include <sstream>
#include "logger.h"
#include "raft_exception.h"
#include "snapshot/send_task.h"

namespace sharkstore {
namespace raft {
namespace impl {

void RaftFsm::becomeLeader() {
    if (state_ == FsmState::kFollower) {
        throw RaftException(
            "[raft->becomeLeader]invalid transition [follower -> leader].");
    }

    step_func_ = std::bind(&RaftFsm::stepLeader, this, std::placeholders::_1);
    reset(term_, true);
    tick_func_ = std::bind(&RaftFsm::tickHeartbeat, this);
    leader_ = node_id_;
    state_ = FsmState::kLeader;

    std::vector<EntryPtr> ents;
    Status s = raft_log_->entries(raft_log_->committed() + 1, kNoLimit, &ents);
    if (!s.ok()) {
        throw RaftException(std::string("[raft->becomeLeader] unexpected error "
                                        "getting uncommitted entries:") +
                            s.ToString());
    }
    int nconf = numOfPendingConf(ents);
    if (nconf > 1) {
        throw RaftException("[raft->becomeLeader]unexpected double uncommitted config "
                            "entry");
    } else if (nconf == 1) {
        pending_conf_ = true;
    }

    // 添加一条空日志 尝试提交前任leader未提交的日志
    EntryPtr entry(new pb::Entry);
    entry->set_type(pb::ENTRY_NORMAL);
    entry->set_term(term_);
    entry->set_index(raft_log_->lastIndex() + 1);
    appendEntry(std::vector<EntryPtr>{entry});

    LOG_INFO("raft[%llu] become leader at term %llu", id_, term_);
}

void RaftFsm::stepLeader(MessagePtr& msg) {
    if (msg->type() == pb::LOCAL_MSG_PROP) {
        if (replicas_.find(node_id_) != replicas_.end() && msg->entries_size() > 0) {
            std::vector<EntryPtr> ents;
            takeEntries(msg, ents);
            // 保证同时只有一个ConfChange
            auto li = raft_log_->lastIndex();
            for (auto& entry : ents) {
                entry->set_index(++li);
                entry->set_term(term_);
                if (entry->type() == pb::ENTRY_CONF_CHANGE) {
                    if (pending_conf_) {
                        // 重新为空日志
                        entry->set_type(pb::ENTRY_NORMAL);
                        entry->clear_data();
                    }
                    pending_conf_ = true;
                }
            }
            appendEntry(ents);
            bcastAppend();
        }
        return;
    }

    // 其他类型的消息需要读取副本的进度
    auto replica = getReplica(msg->from());
    if (replica == nullptr) {
        LOG_WARN("raft[%llu] stepLeader no progress available for %llu", id_,
                 msg->from());
        return;
    }

    Replica& pr = *replica;
    pr.set_active();

    switch (msg->type()) {
        case pb::APPEND_ENTRIES_RESPONSE:
            if (msg->reject()) {
                LOG_DEBUG("raft[%llu] received msgApp "
                          "rejection(lastindex:%llu) from %llu for index %llu",
                          id_, msg->reject_hint(), msg->from(), msg->log_index());

                if (pr.maybeDecrTo(msg->log_index(), msg->reject_hint(), msg->commit())) {
                    if (pr.state() == ReplicaState::kReplicate) {
                        pr.becomeProbe();
                    }
                    sendAppend(msg->from(), pr);
                }
            } else {
                bool old_paused = pr.isPaused();
                if (pr.maybeUpdate(msg->log_index(), msg->commit())) {
                    switch (pr.state()) {
                        case ReplicaState::kProbe:
                            pr.becomeReplicate();
                            break;
                        case ReplicaState::kReplicate:
                            pr.inflight().freeTo(msg->log_index());
                            break;
                        case ReplicaState::kSnapshot:
                            if (pr.needSnapshotAbort()) {
                                LOG_INFO("raft[%llu] snapshot aborted, resumed "
                                         "sending replication to %llu",
                                         id_, msg->from());
                                // 取消快照发送
                                if (sending_snap_ &&
                                    sending_snap_->GetContext().to == msg->from()) {
                                    sending_snap_->Cancel();
                                    sending_snap_.reset();
                                }
                                pr.becomeProbe();
                            }
                            break;
                    }
                    if (maybeCommit()) {
                        bcastAppend();  // commit位置有更新，通知followers
                    } else if (old_paused) {
                        sendAppend(msg->from(), pr);
                    }
                }
            }
            return;

        case pb::HEARTBEAT_RESPONSE:
            pr.resume();
            if (pr.state() == ReplicaState::kReplicate && pr.inflight().full()) {
                pr.inflight().freeFirstOne();
            }
            // 进度没跟上，需要复制
            if (pr.match() < raft_log_->lastIndex() ||
                pr.committed() < raft_log_->committed()) {
                sendAppend(msg->from(), pr);
            }
            return;

        case pb::SNAPSHOT_ACK:
            if (sending_snap_ &&
                sending_snap_->GetContext().uuid == msg->snapshot().uuid()) {
                LOG_DEBUG("raft[%lu] recv snapshot[%lu] ack. seq=%ld, reject=%d", id_,
                          msg->snapshot().uuid(), msg->snapshot().seq(), msg->reject());

                auto s = sending_snap_->RecvAck(msg);
                if (!s.ok()) {
                    LOG_WARN("raft[%lu] ack[seq:%ld, reject:%d, from:%lu] snapshot[%lu] "
                             "failed: %s ",
                             id_, msg->snapshot().seq(), msg->reject(), msg->from(),
                             msg->snapshot().uuid(), s.ToString().c_str());
                }
            }

            return;

        case pb::LOCAL_SNAPSHOT_STATUS:
            if (!sending_snap_ ||
                sending_snap_->GetContext().uuid != msg->snapshot().uuid()) {
                return;
            }
            sending_snap_.reset();

            if (pr.state() != ReplicaState::kSnapshot) {
                return;
            }

            if (msg->reject()) {
                LOG_WARN("raft[%llu] send snapshot to [%llu] failed.", id_, msg->from());
                pr.snapshotFailure();
                pr.becomeProbe();
            } else {
                pr.becomeProbe();
                LOG_WARN("raft[%llu] send snapshot to [%llu] succeed, resumed "
                         "replication [%s]",
                         id_, msg->from(), pr.ToString().c_str());
            }

            pr.pause();
            return;
        default:
            return;
    }
}

void RaftFsm::tickHeartbeat() {
    ++heartbeat_elapsed_;

    if (state_ != FsmState::kLeader) {
        return;
    }

    // 增加副本的inactive_tick
    traverseReplicas([this](uint64_t node, Replica& pr) {
        if (node != node_id_) {
            pr.incr_inactive_tick();
        }
    });

    if (heartbeat_elapsed_ >= sops_.heartbeat_tick) {
        heartbeat_elapsed_ = 0;

        // 检查是否需要提升learner
        if (sops_.auto_promote_learner && !learners_.empty() && !pending_conf_) {
            checkCaughtUp();
        }
    }
}

bool RaftFsm::maybeCommit() {
    std::vector<uint64_t> matches;
    matches.reserve(replicas_.size());
    for (const auto& r : replicas_) {
        matches.push_back(r.second->match());
    }
    std::sort(matches.begin(), matches.end(), std::greater<uint64_t>());

    uint64_t mid = matches[quorum() - 1];
    bool is_commit = raft_log_->maybeCommit(mid, term_);
    if (state_ == FsmState::kLeader) {
        auto it = replicas_.find(node_id_);
        if (it != replicas_.end()) {
            it->second->set_committed(raft_log_->committed());
        }
    }
    return is_commit;
}

void RaftFsm::bcastAppend() {
    traverseReplicas([this](uint64_t node, Replica& pr) {
        if (node != node_id_) this->sendAppend(node, pr);
    });
}

void RaftFsm::sendAppend(uint64_t to, Replica& pr) {
    assert(to == pr.peer().node_id);

    if (pr.isPaused()) {
        return;
    }

    if (pr.inactive_ticks() > sops_.inactive_tick) {
        pr.becomeProbe();
        pr.pause();
        return;
    }

    Status ts, es;
    uint64_t term = 0;
    std::vector<EntryPtr> ents;

    uint64_t fi = raft_log_->firstIndex();
    if (pr.next() >= fi) {
        ts = raft_log_->term(pr.next() - 1, &term);
        es = raft_log_->entries(pr.next(), sops_.max_size_per_msg, &ents);
    }

    // 需要发快照
    if (pr.next() < fi || !ts.ok() || !es.ok()) {
        LOG_INFO("raft[%llu] need snapshot to %llu[next:%llu], fi:%llu, log "
                 "error:%s-%s",
                 id_, to, pr.next(), fi, ts.ToString().c_str(), es.ToString().c_str());

        if (sending_snap_) {
            LOG_WARN("raft[%llu] sendAppend could not send snapshot to %llu(other "
                     "snapshot[%lu] is sending)",
                     id_, to, sending_snap_->GetContext().uuid);
        } else {
            uint64_t snap_index = 0;
            sending_snap_ = newSendSnapTask(to, &snap_index);
            pr.becomeSnapshot(snap_index);
        }
    } else {
        MessagePtr msg(new pb::Message);
        msg->set_type(pb::APPEND_ENTRIES_REQUEST);
        msg->set_to(to);
        msg->set_log_index(pr.next() - 1);  // prev log index
        msg->set_log_term(term);            // prev log term
        msg->set_commit(raft_log_->committed());
        putEntries(msg, ents);

        if (msg->entries_size() > 0) {
            switch (pr.state()) {
                case ReplicaState::kReplicate: {
                    uint64_t last = msg->entries(msg->entries_size() - 1).index();
                    pr.update(last);
                    pr.inflight().add(last);
                    break;
                }
                case ReplicaState::kProbe:
                    pr.pause();
                    break;
                case ReplicaState::kSnapshot:
                    throw RaftException(
                        std::string("[repl->sendAppend][%v] is sending append "
                                    "in unhandled state ") +
                        ReplicateStateName(pr.state()));
            }
        }
        send(msg);
    }
}

void RaftFsm::appendEntry(const std::vector<EntryPtr>& ents) {
    LOG_DEBUG("raft[%llu] append log entry. index: %llu, size: %d", id_, ents[0]->index(),
              ents.size());

    raft_log_->append(ents);
    replicas_[node_id_]->maybeUpdate(raft_log_->lastIndex(), raft_log_->committed());
    maybeCommit();
}

static uint64_t unixNano() {
    auto now = std::chrono::system_clock::now();
    auto count = std::chrono::time_point_cast<std::chrono::nanoseconds>(now).time_since_epoch().count();
    return static_cast<uint64_t>(count);
}

std::shared_ptr<SendSnapTask> RaftFsm::newSendSnapTask(uint64_t to,
                                                       uint64_t* snap_index) {
    SnapContext snap_ctx;
    snap_ctx.id = id_;
    snap_ctx.to = to;
    snap_ctx.from = node_id_;
    snap_ctx.term = term_;
    snap_ctx.uuid = unixNano();

    pb::SnapshotMeta snap_meta;
    // 添加成员信息
    traverseReplicas([&snap_meta](uint64_t node, const Replica& pr) {
        auto p = snap_meta.add_peers();
        auto s = EncodePeer(pr.peer(), p);
        if (!s.ok()) {
            throw RaftException(std::string("create snapshot failed: ") + s.ToString());
        }
    });

    auto snap_data = sm_->GetSnapshot();
    if (snap_data == nullptr) {
        throw RaftException("raft->sendAppend failed to send snapshot, because "
                            "snapshot is unavailable");
    } else if (snap_data->ApplyIndex() < raft_log_->firstIndex() - 1) {
        std::ostringstream ss;
        ss << "raft->sendAppend[" << id_
           << "]failed to send snapshot, because snapshot is invalid(apply="
           << snap_data->ApplyIndex() << ", first=" << raft_log_->firstIndex() << ").";
        throw RaftException(ss.str());
    }

    // set meta index
    snap_meta.set_index(snap_data->ApplyIndex());
    // set meta term
    uint64_t snap_term = 0;
    auto status = raft_log_->term(snap_data->ApplyIndex(), &snap_term);
    if (!status.ok()) {
        std::ostringstream ss;
        ss << "[raft->sendAppend][" << id_ << "] failed to send snapshot to " << to
           << " because snapshot is unavailable, error is: " << status.ToString();
        throw RaftException(ss.str());
    }
    snap_meta.set_term(snap_term);

    // set user context
    std::string context;
    auto s = snap_data->Context(&context);
    if (!s.ok()) {
        throw RaftException(std::string("get snapshot user context failed: ") +
                            s.ToString());
    }
    snap_meta.set_context(std::move(context));

    *snap_index = snap_data->ApplyIndex();

    LOG_DEBUG("raft[%llu] sendAppend [firstindex: %llu, commit: %llu] sent "
              "snapshot[index: %llu, term: %llu] to [%llu]",
              id_, raft_log_->firstIndex(), raft_log_->committed(),
              snap_data->ApplyIndex(), snap_term, to);

    return std::make_shared<SendSnapTask>(snap_ctx, std::move(snap_meta), snap_data);
}

void RaftFsm::checkCaughtUp() {
    Peer max_peer;
    uint64_t max_match = 0;
    for (const auto& p : learners_) {
        const auto& pr = *p.second;
        if (pr.match() >= max_match) {
            max_match = pr.match();
            max_peer = pr.peer();
        }
    }

    auto lasti = raft_log_->lastIndex();
    auto precent_threshold = (sops_.promote_gap_percent * lasti) / 100;
    uint64_t final_threshold = std::max(sops_.promote_gap_threshold, precent_threshold);

    assert(lasti >= max_match);
    if (lasti - max_match < final_threshold) {
        LOG_INFO("raft[%lu] start promote learner %s [match:%lu, lasti:%lu] at term %lu.", id_,
                 max_peer.ToString().c_str(), max_match, lasti, term_);

        max_peer.type = PeerType::kNormal;
        ConfChange cc;
        cc.type = ConfChangeType::kPromote;
        cc.peer = max_peer;
        std::string str;
        auto s = EncodeConfChange(cc, &str);
        if (!s.ok()) {
            throw RaftException(std::string("promote caughtup learner failed:") +
                                s.ToString());
        }

        auto msg = std::make_shared<pb::Message>();
        msg->set_type(pb::LOCAL_MSG_PROP);
        auto entry = msg->add_entries();
        entry->set_type(pb::ENTRY_CONF_CHANGE);
        entry->mutable_data()->swap(str);
        Step(msg);
    }
}

} /* namespace impl */
} /* namespace raft */
} /* namespace sharkstore */
